File: ios/Dash/Core/Contracts/ConversationContracts.swift
import Foundation

enum ConversationStatus: String, Codable, Hashable, Sendable {
  case idle
  case running
  case interrupted
  case archived
  case deleted
}

enum MessageRole: String, Codable, Hashable, Sendable {
  case user
  case assistant
}

enum MessageStatus: String, Codable, Hashable, Sendable {
  case accepted
  case streaming
  case completed
  case cancelled
  case failed
  case interrupted
}

enum ImageMediaType: String, Codable, Hashable, Sendable {
  case jpeg = "image/jpeg"
  case png = "image/png"
  case gif = "image/gif"
  case webp = "image/webp"
}

struct MessageImage: Codable, Hashable, Sendable {
  let mediaType: ImageMediaType
  let data: String
}

/// What a `notice` message is reporting. Unknown values decode to `.unknown`
/// rather than throwing, so a gateway that learns a new notice kind does not
/// break decoding on an app that predates it.
enum NoticeKind: String, Codable, Hashable, Sendable {
  case skillLearned = "skill_learned"
  case memorySaved = "memory_saved"
  case unknown

  init(from decoder: Decoder) throws {
    let raw = try decoder.singleValueContainer().decode(String.self)
    self = NoticeKind(rawValue: raw) ?? .unknown
  }
}

enum MessageContent: Codable, Hashable, Sendable {
  case user(text: String, images: [MessageImage]?)
  case assistant(events: [AgentEvent])
  /// Something the gateway recorded after a turn finished — a skill learned or
  /// a memory saved by its post-turn review.
  case notice(kind: NoticeKind, text: String)
  /// A content type this build does not know about. Same reasoning as
  /// `AgentEvent.unknown`: a whole page of messages decodes as a unit, so a
  /// single unrecognised message must degrade rather than fail the page and
  /// blank the transcript.
  case unknown(type: String)

  private enum CodingKeys: String, CodingKey {
    case type
    case text
    case images
    case events
    case kind
  }

  init(from decoder: Decoder) throws {
    let container = try decoder.container(keyedBy: CodingKeys.self)
    let type = try container.decode(String.self, forKey: .type)
    switch type {
    case "user":
      self = .user(
        text: try container.decode(String.self, forKey: .text),
        images: try container.decodeIfPresent([MessageImage].self, forKey: .images)
      )
    case "assistant":
      self = .assistant(events: try container.decode([AgentEvent].self, forKey: .events))
    case "notice":
      self = .notice(
        kind: try container.decode(NoticeKind.self, forKey: .kind),
        text: try container.decode(String.self, forKey: .text)
      )
    default:
      self = .unknown(type: type)
    }
  }

  func encode(to encoder: Encoder) throws {
    var container = encoder.container(keyedBy: CodingKeys.self)
    switch self {
    case let .user(text, images):
      try container.encode("user", forKey: .type)
      try container.encode(text, forKey: .text)
      try container.encodeIfPresent(images, forKey: .images)
    case let .assistant(events):
      try container.encode("assistant", forKey: .type)
      try container.encode(events, forKey: .events)
    case let .notice(kind, text):
      try container.encode("notice", forKey: .type)
      try container.encode(kind, forKey: .kind)
      try container.encode(text, forKey: .text)
    case let .unknown(type):
      try container.encode(type, forKey: .type)
    }
  }
}

struct ConversationSummaryDTO: Codable, Hashable, Identifiable, Sendable {
  let id: String
  let agentId: String
  let agentName: String
  let title: String
  let revision: Int
  let status: ConversationStatus
  let activeTurnId: String?
  let owningIssueId: String?
  let projectId: String?
  let lastSeq: Int
  let lastMessagePreview: String?
  let createdAt: Date
  let updatedAt: Date
  let deletedAt: Date?
}

struct ConversationMessageDTO: Codable, Hashable, Identifiable, Sendable {
  let id: String
  let conversationId: String
  let turnId: String
  let ordinal: Int
  let role: MessageRole
  let status: MessageStatus
  let content: MessageContent
  let createdAt: Date
  let updatedAt: Date
}

struct ConversationPageDTO: Codable, Hashable, Sendable {
  let items: [ConversationSummaryDTO]
  let nextCursor: String?
}

struct ConversationMessagePageDTO: Codable, Hashable, Sendable {
  let items: [ConversationMessageDTO]
  let nextCursor: String?
  let throughSeq: Int
}

struct CreateConversationRequest: Codable, Hashable, Sendable {
  let agentId: String
  let requestId: String
  let title: String?
  let owningIssueId: String?
  let projectId: String?
}

enum NullablePatchField<Value: Hashable & Sendable>: Hashable, Sendable {
  case omitted
  case value(Value)
  case null
}

enum PatchConversationRequestError: Error, Equatable, Sendable {
  case emptyPatch
}

struct PatchConversationRequest: Codable, Hashable, Sendable {
  let title: String?
  let owningIssueId: String?
  let projectId: String?

  private let includedKeys: Set<String>

  init(
    title: String? = nil,
    owningIssueId: NullablePatchField<String> = .omitted,
    projectId: NullablePatchField<String> = .omitted
  ) throws {
    self.title = title
    self.owningIssueId = if case let .value(value) = owningIssueId { value } else { nil }
    self.projectId = if case let .value(value) = projectId { value } else { nil }

    var keys: Set<String> = []
    if title != nil { keys.insert(CodingKeys.title.rawValue) }
    if owningIssueId != .omitted { keys.insert(CodingKeys.owningIssueId.rawValue) }
    if projectId != .omitted { keys.insert(CodingKeys.projectId.rawValue) }
    guard keys.isEmpty == false else { throw PatchConversationRequestError.emptyPatch }
    includedKeys = keys
  }

  private enum CodingKeys: String, CodingKey {
    case title
    case owningIssueId
    case projectId
  }

  init(from decoder: Decoder) throws {
    let container = try decoder.container(keyedBy: CodingKeys.self)
    title = try container.decodeIfPresent(String.self, forKey: .title)
    owningIssueId = try container.decodeIfPresent(String.self, forKey: .owningIssueId)
    projectId = try container.decodeIfPresent(String.self, forKey: .projectId)
    includedKeys = Set(container.allKeys.map(\.rawValue))
  }

  func encode(to encoder: Encoder) throws {
    guard includedKeys.isEmpty == false else { throw PatchConversationRequestError.emptyPatch }
    var container = encoder.container(keyedBy: CodingKeys.self)
    if includedKeys.contains(CodingKeys.title.rawValue) {
      try container.encodeIfPresent(title, forKey: .title)
      if title == nil { try container.encodeNil(forKey: .title) }
    }
    if includedKeys.contains(CodingKeys.owningIssueId.rawValue) {
      try container.encodeIfPresent(owningIssueId, forKey: .owningIssueId)
      if owningIssueId == nil { try container.encodeNil(forKey: .owningIssueId) }
    }
    if includedKeys.contains(CodingKeys.projectId.rawValue) {
      try container.encodeIfPresent(projectId, forKey: .projectId)
      if projectId == nil { try container.encodeNil(forKey: .projectId) }
    }
  }
}

enum TurnOutcome: String, Codable, Hashable, Sendable {
  case completed
  case cancelled
}

enum ReplayPayload: Codable, Hashable, Sendable {
  case accepted(userMessageId: String, assistantMessageId: String, revision: Int)
  case event(event: AgentEvent)
  case done(outcome: TurnOutcome?)
  case error(error: String, code: String?, retryable: Bool?)

  private enum CodingKeys: String, CodingKey {
    case type
    case userMessageId
    case assistantMessageId
    case revision
    case event
    case outcome
    case error
    case code
    case retryable
  }

  private enum Kind: String, Codable {
    case accepted
    case event
    case done
    case error
  }

  init(from decoder: Decoder) throws {
    let container = try decoder.container(keyedBy: CodingKeys.self)
    switch try container.decode(Kind.self, forKey: .type) {
    case .accepted:
      self = .accepted(
        userMessageId: try container.decode(String.self, forKey: .userMessageId),
        assistantMessageId: try container.decode(String.self, forKey: .assistantMessageId),
        revision: try container.decode(Int.self, forKey: .revision)
      )
    case .event:
      self = .event(event: try container.decode(AgentEvent.self, forKey: .event))
    case .done:
      self = .done(outcome: try container.decodeIfPresent(TurnOutcome.self, forKey: .outcome))
    case .error:
      self = .error(
        error: try container.decode(String.self, forKey: .error),
        code: try container.decodeIfPresent(String.self, forKey: .code),
        retryable: try container.decodeIfPresent(Bool.self, forKey: .retryable)
      )
    }
  }

  func encode(to encoder: Encoder) throws {
    var container = encoder.container(keyedBy: CodingKeys.self)
    switch self {
    case let .accepted(userMessageId, assistantMessageId, revision):
      try container.encode(Kind.accepted, forKey: .type)
      try container.encode(userMessageId, forKey: .userMessageId)
      try container.encode(assistantMessageId, forKey: .assistantMessageId)
      try container.encode(revision, forKey: .revision)
    case let .event(event):
      try container.encode(Kind.event, forKey: .type)
      try container.encode(event, forKey: .event)
    case let .done(outcome):
      try container.encode(Kind.done, forKey: .type)
      try container.encodeIfPresent(outcome, forKey: .outcome)
    case let .error(error, code, retryable):
      try container.encode(Kind.error, forKey: .type)
      try container.encode(error, forKey: .error)
      try container.encodeIfPresent(code, forKey: .code)
      try container.encodeIfPresent(retryable, forKey: .retryable)
    }
  }
}

struct ReplayEntryDTO: Codable, Hashable, Sendable {
  let seq: Int
  let msgId: String
  let agentId: String
  let conversationId: String
  let timestamp: Date
  let payload: ReplayPayload
}

struct ReplayPageDTO: Codable, Hashable, Sendable {
  let entries: [ReplayEntryDTO]
}

File: android/core/contracts/src/main/kotlin/app/dash/core/contracts/ConversationContracts.kt
@file:UseSerializers(JsonIntegralLongSerializer::class)

package app.dash.core.contracts

import kotlinx.serialization.KSerializer
import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable
import kotlinx.serialization.SerializationException
import kotlinx.serialization.UseSerializers
import kotlinx.serialization.descriptors.PrimitiveKind
import kotlinx.serialization.descriptors.PrimitiveSerialDescriptor
import kotlinx.serialization.descriptors.buildClassSerialDescriptor
import kotlinx.serialization.encoding.Decoder
import kotlinx.serialization.encoding.Encoder
import kotlinx.serialization.json.JsonNull
import kotlinx.serialization.json.JsonObject
import kotlinx.serialization.json.JsonPrimitive
import kotlinx.serialization.json.buildJsonObject
import kotlinx.serialization.json.decodeFromJsonElement
import kotlinx.serialization.json.encodeToJsonElement
import kotlinx.serialization.json.put

@Serializable
enum class ConversationStatus {
  @SerialName("idle")
  Idle,

  @SerialName("running")
  Running,

  @SerialName("interrupted")
  Interrupted,

  @SerialName("archived")
  Archived,

  @SerialName("deleted")
  Deleted,
}

@Serializable
enum class ConversationMessageStatus {
  @SerialName("accepted")
  Accepted,

  @SerialName("streaming")
  Streaming,

  @SerialName("completed")
  Completed,

  @SerialName("cancelled")
  Cancelled,

  @SerialName("failed")
  Failed,

  @SerialName("interrupted")
  Interrupted,
}

@Serializable
enum class ConversationRole {
  @SerialName("user")
  User,

  @SerialName("assistant")
  Assistant,
}

@Serializable
data class MobileImage(val mediaType: String, val data: String)

@Serializable(with = ConversationNoticeKindSerializer::class)
sealed interface ConversationNoticeKind {
  data object SkillLearned : ConversationNoticeKind

  data object MemorySaved : ConversationNoticeKind

  data class Unknown(val wireValue: String) : ConversationNoticeKind
}

object ConversationNoticeKindSerializer : KSerializer<ConversationNoticeKind> {
  override val descriptor = PrimitiveSerialDescriptor("ConversationNoticeKind", PrimitiveKind.STRING)

  override fun deserialize(decoder: Decoder): ConversationNoticeKind = when (val value = decoder.decodeString()) {
    "skill_learned" -> ConversationNoticeKind.SkillLearned
    "memory_saved" -> ConversationNoticeKind.MemorySaved
    else -> {
      if (value.isEmpty()) throw SerializationException("Conversation notice kind must be nonempty")
      ConversationNoticeKind.Unknown(value)
    }
  }

  override fun serialize(encoder: Encoder, value: ConversationNoticeKind) {
    val wireValue = when (value) {
      ConversationNoticeKind.SkillLearned -> "skill_learned"
      ConversationNoticeKind.MemorySaved -> "memory_saved"
      is ConversationNoticeKind.Unknown -> value.wireValue
    }
    if (wireValue.isEmpty()) throw SerializationException("Conversation notice kind must be nonempty")
    encoder.encodeString(wireValue)
  }
}

@Serializable(with = ConversationContentSerializer::class)
sealed interface ConversationContent {
  @Serializable
  data class User(val text: String, val images: List<MobileImage> = emptyList()) : ConversationContent

  @Serializable
  data class Assistant(val events: List<AgentEvent>) : ConversationContent

  @Serializable
  data class Notice(val kind: ConversationNoticeKind, val text: String) : ConversationContent

  data class Unknown(val type: String, val raw: JsonObject) : ConversationContent
}

object ConversationContentSerializer : KSerializer<ConversationContent> {
  override val descriptor = buildClassSerialDescriptor("ConversationContent")

  override fun deserialize(decoder: Decoder): ConversationContent {
    val raw = decoder.decodeSerializableValue(JsonObject.serializer())
    val type = raw["type"] as? JsonPrimitive
    if (type == null || !type.isString || type.content.isEmpty()) {
      throw SerializationException("ConversationContent.type must be a nonempty string")
    }
    return when (type.content) {
      "user" -> {
        raw.requireExactContentKeys(setOf("type", "text"), setOf("images"))
        ContractJson.strict.decodeFromJsonElement<ConversationContent.User>(JsonObject(raw - "type"))
      }
      "assistant" -> {
        raw.requireExactContentKeys(setOf("type", "events"))
        ContractJson.strict.decodeFromJsonElement<ConversationContent.Assistant>(JsonObject(raw - "type"))
      }
      "notice" -> {
        raw.requireExactContentKeys(setOf("type", "kind", "text"))
        ContractJson.strict.decodeFromJsonElement<ConversationContent.Notice>(JsonObject(raw - "type"))
      }
      else -> ConversationContent.Unknown(type.content, raw)
    }
  }

  override fun serialize(encoder: Encoder, value: ConversationContent) {
    val raw = when (value) {
      is ConversationContent.User -> taggedContent("user", value)
      is ConversationContent.Assistant -> taggedContent("assistant", value)
      is ConversationContent.Notice -> taggedContent("notice", value)
      is ConversationContent.Unknown -> {
        val rawType = value.raw["type"] as? JsonPrimitive
        if (
          value.type.isEmpty() ||
          rawType == null ||
          !rawType.isString ||
          rawType.content != value.type
        ) {
          throw SerializationException("ConversationContent.type must match its raw object")
        }
        value.raw
      }
    }
    encoder.encodeSerializableValue(JsonObject.serializer(), raw)
  }

  private inline fun <reified T> taggedContent(type: String, value: T): JsonObject = buildJsonObject {
    put("type", type)
    ContractJson.strict.encodeToJsonElement(value).let { encoded ->
      (encoded as JsonObject).forEach { (key, node) -> put(key, node) }
    }
  }

  private fun JsonObject.requireExactContentKeys(
    required: Set<String>,
    optional: Set<String> = emptySet(),
  ) {
    if (!keys.containsAll(required) || keys.any { it !in required && it !in optional }) {
      throw SerializationException("ConversationContent has invalid keys")
    }
  }
}

@Serializable
data class ConversationDefaults(val defaultConversationTitle: String) {
  init {
    require(defaultConversationTitle == DEFAULT_CONVERSATION_TITLE) {
      "ConversationDefaults.defaultConversationTitle must be New Conversation"
    }
  }

  companion object {
    const val DEFAULT_CONVERSATION_TITLE = "New Conversation"
  }
}

@Serializable
data class ConversationSummary(
  val id: String,
  val agentId: String,
  val agentName: String,
  val title: String,
  val revision: Long,
  val status: ConversationStatus,
  val activeTurnId: String?,
  val owningIssueId: String?,
  val projectId: String?,
  val lastSeq: Long,
  val lastMessagePreview: String?,
  val createdAt: String,
  val updatedAt: String,
  val deletedAt: String? = null,
)

@Serializable
data class ConversationMessage(
  val id: String,
  val conversationId: String,
  val turnId: String,
  val ordinal: Long,
  val role: ConversationRole,
  val status: ConversationMessageStatus,
  val content: ConversationContent,
  val createdAt: String,
  val updatedAt: String,
)

@Serializable
data class ConversationPage(val items: List<ConversationSummary>, val nextCursor: String?)

@Serializable
data class ConversationMessagePage(
  val items: List<ConversationMessage>,
  val nextCursor: String?,
  val throughSeq: Long,
)

@Serializable
data class ConversationCreateRequest(
  val agentId: String,
  val requestId: String,
  val title: String? = null,
  val owningIssueId: String? = null,
  val projectId: String? = null,
)

sealed interface NullablePatchField<out T> {
  data object Omitted : NullablePatchField<Nothing>

  data class Value<T>(val value: T) : NullablePatchField<T>

  data object Null : NullablePatchField<Nothing>
}

@Serializable(with = ConversationPatchRequestSerializer::class)
data class ConversationPatchRequest(
  val title: String? = null,
  val owningIssueId: NullablePatchField<String> = NullablePatchField.Omitted,
  val projectId: NullablePatchField<String> = NullablePatchField.Omitted,
)

object ConversationPatchRequestSerializer : KSerializer<ConversationPatchRequest> {
  override val descriptor = buildClassSerialDescriptor("ConversationPatchRequest")

  override fun deserialize(decoder: Decoder): ConversationPatchRequest {
    val raw = decoder.decodeSerializableValue(JsonObject.serializer())
    if (raw.isEmpty()) throw SerializationException("ConversationPatchRequest must not be empty")
    if (raw.keys.any { it !in setOf("title", "owningIssueId", "projectId") }) {
      throw SerializationException("ConversationPatchRequest has invalid keys")
    }
    val title = when (val node = raw["title"]) {
      null -> null
      JsonNull -> throw SerializationException("ConversationPatchRequest.title must not be null")
      is JsonPrimitive -> if (node.isString) node.content else {
        throw SerializationException("ConversationPatchRequest.title must be a string")
      }
      else -> throw SerializationException("ConversationPatchRequest.title must be a string")
    }
    return ConversationPatchRequest(
      title = title,
      owningIssueId = raw.patchField("owningIssueId"),
      projectId = raw.patchField("projectId"),
    )
  }

  override fun serialize(encoder: Encoder, value: ConversationPatchRequest) {
    val raw = buildJsonObject {
      value.title?.let { put("title", it) }
      putPatchField("owningIssueId", value.owningIssueId)
      putPatchField("projectId", value.projectId)
    }
    if (raw.isEmpty()) throw SerializationException("ConversationPatchRequest must not be empty")
    encoder.encodeSerializableValue(JsonObject.serializer(), raw)
  }

  private fun JsonObject.patchField(field: String): NullablePatchField<String> = when (
    val node = get(field)
  ) {
    null -> NullablePatchField.Omitted
    JsonNull -> NullablePatchField.Null
    is JsonPrimitive -> if (node.isString) NullablePatchField.Value(node.content) else {
      throw SerializationException("ConversationPatchRequest.$field must be a string or null")
    }
    else -> throw SerializationException("ConversationPatchRequest.$field must be a string or null")
  }

  private fun kotlinx.serialization.json.JsonObjectBuilder.putPatchField(
    field: String,
    value: NullablePatchField<String>,
  ) {
    when (value) {
      NullablePatchField.Omitted -> Unit
      NullablePatchField.Null -> put(field, JsonNull)
      is NullablePatchField.Value -> put(field, value.value)
    }
  }
}
